api/routes/auth.js

```diff
@@ -0,0 +1,149 @@
+const router = require('express').Router();
+const User = require('../models/user');
+const bcrypt = require('bcrypt');
+const jsonwebtoken = require('jsonwebtoken');
+
+const { SECRET_KEY } = process.env;
+
+  /***
+ * - [ ] **Create a `POST /api/signup` route:** Create a new route that allows someone to create an account. Securely store the password using the `bcrypt` package. 
+ * On successful creation, return a JWT token. You should return an error in the following cases:
+  * Username is not provided
+  * Username is already taken
+  * Password is not provided
+  * Password is less than 8 characters
+ */
+
+router.post('/signup', async (req, res, next) => {
+  const status = 201;
+  try {
+    const {username, password} = req.body;
+    if (!username) {
+        throw new Error (`User name not provided. Please update your request and retry.`);
+    }
+    if (!password) {
+        throw new Error (`Password not provided. Please update your request and retry.`);
+    }
+    if (password.length < 8) {
+        throw new Error (`Password does not meet minimum length requirements. Please update your request and retry.`)
+    }
+    const user = await User.findOne({username});
+    if (user) {
+        throw new Error (`User ${username} already exists.`)
+    }
+    const saltRounds = 10;
+    const hashedPassword = await bcrypt.hash(password, saltRounds);
+    const newUser = await User.create({
+        username,
+        password: hashedPassword
+    });   
+    const payload = {username: newUser.username, password: newUser.password };
+    const options = { expiresIn: '1 day' };
+    const token = jsonwebtoken.sign(payload, SECRET_KEY, options);
+    res.status(201).json({ status, token});
+  } catch (e) {
+      console.error(e);
+      const error = new Error(`An error has occurred. Please retry your request.`);
+      error.status = 400;
+      next(error);
+  }
+  
+});
+
+/**
+ * 
+ * - [ ] **Create a `POST /api/login` route:** Create a new route that allows someone to login. On successful creation, return a JWT token. You should return an error in the following cases:
+  * Username is not found
+  * Username and password do not match
+ * 
+ */
+
+router.post('/login', async (req, res, next) => {
+    const status = 201;
+    try {
+        const {username, password} = req.body;
+        const user = await User.findOne({username});
+        if (!user) {
+            const error = new Error(`Username could not be found.`);
+            error.message = 404;
+            return next(error);
+        }
+        const isValid = await bcrypt.compare(password, user.password);
+        if (!isValid) {
+            const error = new Error(`Password is invalid.`);

```

Comment:
You'll want to give errors that are vaguer in the event of logging in. Otherwise, it's easier for someone to "guess" their way into someone else's account.